File: test/pgo/ExpressionCodeGenRunFailureTest.java
package pgo;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import pgo.model.tla.TLAExpression;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static pgo.model.tla.TLABuilder.*;

@RunWith(Parameterized.class)
public class ExpressionCodeGenRunFailureTest {
	@Parameters
	public static List<Object[]> data(){
		return Arrays.asList(new Object[][] {
			// if expression tests
			{
					caseexp(arms(arm(bool(false), str("Hello world"))), null),
					Collections.emptyList(),
					Collections.singletonList("panic: No matching case!"),
			},
		});
	}

	private TLAExpression result;
	private List<IntegrationTestingUtils.KeyValue> vars;
	private List<String> expected;

	public ExpressionCodeGenRunFailureTest(TLAExpression result, List<IntegrationTestingUtils.KeyValue> vars, List<String> expected) {
		this.result = result;
		this.vars = vars;
		this.expected = expected;
	}

	@Test
	public void test() throws IOException {
		// try to run the compiled Go code, check that it prints the right thing
		IntegrationTestingUtils.testCompileExpression(result, vars, compiledOutputPath ->
			IntegrationTestingUtils.testRunGoCodeShouldPanic(compiledOutputPath, expected));
	}
}

File: test/pgo/ExpressionCodeGenRunFailureTest.scala
package pgo

import org.scalactic.source.Position
import org.scalatest.funsuite.AnyFunSuite
import pgo.model.tla.TLAExpression
import pgo.model.tla.TLABuilder._

import scala.jdk.CollectionConverters._

class ExpressionCodeGenRunFailureTest extends AnyFunSuite {
  def check(tag: String)(expr: TLAExpression, vars: List[IntegrationTestingUtils.KeyValue], expected: String)(implicit pos: Position): Unit =
    test(tag) {
      IntegrationTestingUtils.testCompileExpression(expr, vars.asJava, { outputPath =>
        IntegrationTestingUtils.testRunGoCodeShouldPanic(outputPath, expected.linesIterator.toList.asJava)
      })
    }

  check("unsatisfiable case expr")(
    expr=caseexp(arms(arm(bool(false), str("Hello world"))), null),
    vars=Nil,
    "panic: No matching case!")
}